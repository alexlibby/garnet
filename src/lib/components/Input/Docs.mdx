import { Canvas, Meta } from "@storybook/blocks";
import { ArgTypes } from "@storybook/blocks";
import * as InputStories from "./Input.stories";

<Meta of={InputStories} />

# Input

Input is the primary component. It has four possible states.

- [Default](#text)
- [No Label](#nolabel)
- [Disabled Input](#disabledinput)

## Text

This is text for our text input field component

<Canvas of={InputStories.Default} />

## NoLabel

This is to set or hide the label for the input field component

<Canvas of={InputStories.NoLabel} />

## DisabledInput

This disables the component

<Canvas of={InputStories.Disabled} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={InputStories} />
