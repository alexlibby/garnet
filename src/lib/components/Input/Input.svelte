<svelte:options customElement="garnet-input" />

<script>
	import { createEventDispatcher } from 'svelte';

	export let label = "Label:";
	export let placeholder = "default placeholder";
	export let disabled = false;
	export let inputName = "";
	export let fieldID = "";
	export let value = undefined;

	const dispatch = createEventDispatcher();
  
	const onInput =(e)=>{
		dispatch('input', { text: e.target.value});
	}
</script>

<div class="garnet-input">
	{#if label}
		<label for={fieldID}>
			{label}
		</label>
	{/if}
	<input
		type="text"
		id={fieldID}
		name={inputName}
		{placeholder}
		{disabled}
		bind:value
		on:input={onInput}
	/>
</div>

<style>
  .garnet-input {
		display: flex;
		flex-direction: row;
		font-family: Arial, Helvetica, sans-serif;
	}

	input[type="text"] {
		width: 200px;
		border-radius: 4px;
		border-color: #733635;
		height: 30px;
		outline: none;
	}

	label {
		padding-right: 10px;
		display: flex;
		align-self: center;
	}

	input:disabled {
		cursor: not-allowed;
	}
</style>
