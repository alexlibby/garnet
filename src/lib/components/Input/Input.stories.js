import Input from "./Input.svelte";
import { action } from "@storybook/addon-actions";
import { fn } from "@storybook/test";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/Basic Components/Input",
  component: Input,
  argTypes: {
    disabled: { control: "boolean" },
    placeholder: "",
  },
  props: {
    label: "Text:",
    placeholder: "Enter your text here",
    disabled: false,
  },
  on: { input: fn().mockName("on-input") },
};

export const Default = (args) => ({
  Component: Input,
  ...args,
  props: {},
  on: { input: fn().mockName("on-input") },
});

Default.parameters = {
  badges: [BADGES.ALPHA],
};

export const NoLabel = (args) => ({
  Component: Input,
  ...args,
  props: {
    label: "",
  },
  on: { input: fn().mockName("on-input") },
});

NoLabel.parameters = {
  badges: [BADGES.ALPHA],
};

export const Disabled = (args) => ({
  Component: Input,
  ...args,
  props: {
    label: "Text:",
    disabled: true,
    placeholder: "disabled field",
  },
});

Disabled.parameters = {
  badges: [BADGES.ALPHA],
};
