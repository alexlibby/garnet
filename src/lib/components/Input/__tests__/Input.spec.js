import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import Input from "../../Input/Input.svelte";

const InputHTML = `<div class="garnet s--FAEzmeZd6RB"><label for="" class="s--FAEzmeZd6RB">Label:</label> <input type="text" id="" name="" placeholder="default placeholder" class="s--FAEzmeZd6RB"></div>`;
expect.extend(AxeMatchers);

describe("Tests for Input", () => {
  const props = {
    fieldID: "Input",
    disabled: false,
    placeholder: "example text",
  };

  test("should render properly", () => {
    const result = render(Input, props);
    expect(() => result).not.toThrow();
  });

  test("should show a textbox with correct placeholder text", () => {
    render(Input, props);
    const inputPlaceholder = screen.getByRole("textbox").placeholder;
    expect(inputPlaceholder).toEqual("example text");
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Input({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => InputHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
