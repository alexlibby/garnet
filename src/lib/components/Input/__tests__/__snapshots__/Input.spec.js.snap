// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Input > get a snapshot of component 1`] = `"<div class="garnet-input s--FAEzmeZd6RB"><label for="Input" class="s--FAEzmeZd6RB">Label:</label> <input type="text" id="Input" name="" placeholder="example text" class="s--FAEzmeZd6RB"></div><!--<Input>-->"`;
